internal/pkg/aws/s3/s3.go

```diff
@@ -52,3 +61,64 @@ func (s *Service) PutArtifact(bucket, fileName string, data io.Reader) (string,
 
 	return resp.Location, nil
 }
+
+// EmptyBucket deletes all objects within the bucket.
+func (s *Service) EmptyBucket(bucket string) error {
+	var listResp *s3.ListObjectVersionsOutput
+	var err error
+	listParams := &s3.ListObjectVersionsInput{
+		Bucket: aws.String(bucket),
+	}
+	// Remove all versions of all objects.
+	for {
+		listResp, err = s.s3Svc.ListObjectVersions(listParams)
+		if err != nil {
+			return fmt.Errorf("list objects for bucket %s: %w", bucket, err)
+		}
+		if len(listResp.Versions) == 0 {
+			break
+		}
+		objectsID := make([]*s3.ObjectIdentifier, 0, batchSize)
+		for _, object := range listResp.Versions {
+			objectsID = append(objectsID, &s3.ObjectIdentifier{
+				Key:       object.Key,
+				VersionId: object.VersionId,
+			})
+		}
+		_, err = s.s3Svc.DeleteObjects(&s3.DeleteObjectsInput{
+			Bucket: aws.String(bucket),
+			Delete: &s3.Delete{
+				Objects: objectsID,
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("delete objects for bucket %s: %w", bucket, err)
+		}
+		if aws.BoolValue(listResp.IsTruncated) {
+			listParams.KeyMarker = listResp.NextKeyMarker
+			listParams.VersionIdMarker = listResp.NextVersionIdMarker
+		}
+	}
+	// After deleting all versions for all objects, remove all delete marker.
+	// For info on "delete marker": https://docs.aws.amazon.com/AmazonS3/latest/dev/DeleteMarker.html
+	if len(listResp.DeleteMarkers) == 0 {
+		return nil
+	}
+	var deleteMarkers []*s3.ObjectIdentifier
+	for _, deleteMarker := range listResp.DeleteMarkers {
+		deleteMarkers = append(deleteMarkers, &s3.ObjectIdentifier{
+			Key:       deleteMarker.Key,
+			VersionId: deleteMarker.VersionId,
+		})
+	}
+	_, err = s.s3Svc.DeleteObjects(&s3.DeleteObjectsInput{
+		Bucket: aws.String(bucket),
+		Delete: &s3.Delete{
+			Objects: deleteMarkers,
+		},
+	})
```

Comment:
Is it possible that we have more than 1000 deletemarkers and we need to split these up to multiple requests?

Reply:
Good questions! I am not sure either and they don't have a `nextdeletemarker` such kind of stuff, nor do they specify there could be some limitation for the length of "DeleteMarkers" field.